sycl/source/detail/pi.cpp

```diff
@@ -74,18 +78,21 @@ void *loadPlugin(const std::string &PluginPath) {
 // needs to setup infrastructure to route PI_CALLs to the appropriate plugins.
 // Currently, we bind to a singe plugin.
 bool bindPlugin(void *Library) {
-#define STRINGIZE(x) #x
-
-#define _PI_API(api)                                                           \
-  decltype(&api) api##_ptr = ((decltype(&api))(                                \
-      getOsLibraryFuncAddress(Library, STRINGIZE(api##OclPtr))));              \
-  if (!api##_ptr)                                                              \
-    return false;                                                              \
-  api = *api##_ptr;
-#include <CL/sycl/detail/pi.def>
-
-#undef STRINGIZE
-#undef _PI_API
+
+  decltype(::piPluginInit) *PluginInitializeFunction = (decltype(
+      &::piPluginInit))(getOsLibraryFuncAddress(Library, "piPluginInit"));
+  int err = PluginInitializeFunction(&PluginInformation);
+  int CompareVersions =
+      strcmp(PluginInformation.PiVersion, PluginInformation.PluginVersion);
+
+  // CompareVersions >= 0, Plugin Interface supports same/higher PI version as
+  // the Plugin.
+  // TODO: When Plugin supports lower version of PI, check for backward
+  // compatibility.
+  assert((CompareVersions >= 0) && "Plugin Interface supports lower PI version "
+                                   "than Plugin. Update library.");
+  // Reaching here means CompareVersions>=0, make sure err is PI_SUCCESS.
+  assert((err == PI_SUCCESS) && "Unexpected error when binding to Plugin.");
   return true;
```

Comment:
We need a way to report errors other than `true`/`false` to enable the caller to report/deal with the specific problems when loading plugins.

Reply:
Yes. Let me add a TODO.